src/circular_buffer.rs

```diff
@@ -0,0 +1,313 @@
+    /// assert_eq!(buf.index(), 0);
+    /// ```
+    #[inline]
+    pub fn index(&self) -> usize {
+        self.index
+    }
+
+    /// Returns the fist adressable index in the buffer.
+    #[inline]
+    pub fn first_index(&self) -> usize {
+        let index = self.index() as isize;
+        let len = self.len() as isize;
+        let capacity = self.capacity() as isize;
+        ((index - len + capacity) % capacity) as usize
+    }
+
+    /// Returns `true` if the buffer is empty.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let buf = CircularBuffer::new([0;8]);
+    /// assert!(buf.is_empty());
+    /// ```
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns `true` if the buffer is saturated.
+    ///
+    /// Note that elements can still be inserted but will overwrite
+    /// the last ones.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let buf = CircularBuffer::from_array([1,2,3]);
+    /// assert!(buf.is_full());
+    /// ```
+    #[inline]
+    pub fn is_full(&self) -> bool {
+        self.len() == self.capacity()
+    }
+
+    /// Pushes `element` into the buffer.
+    ///
+    /// This will overwrite an old value if the buffer is full.
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let mut buf = CircularBuffer::new([0; 2]);
+    /// buf.push(1);
+    /// buf.push(2);
+    /// buf.push(3);
+    /// assert_eq!(buf.pop(), Some(3));
+    /// assert_eq!(buf.pop(), Some(2));
+    /// assert_eq!(buf.pop(), None);
+    /// ```
+    #[inline]
+    pub fn push(&mut self, element: T) {
+        let len = self.len();
+        let capacity = self.capacity();
+        let index = self.index();
+        if len < capacity {
+            unsafe {
+                self.set_len(len + 1);
+            }
+        }
+        {
+            let slice = self.array.as_mut();
+            slice[index] = element;
+        }
+        unsafe {
+            self.set_index((index + 1) as isize);
+        }
+    }
+
+    /// Pops the most recently inserted element.
+    ///
+    /// Returns `None` if the buffer is empty.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let mut buf = CircularBuffer::from_array([1,2]);
+    /// assert_eq!(buf.pop(), Some(2));
+    /// assert_eq!(buf.pop(), Some(1));
+    /// assert_eq!(buf.pop(), None);
+    /// ```
+    #[inline]
+    pub fn pop(&mut self) -> Option<T> {
+        if self.is_empty() {
+            return None;
+        }
+        let len = self.len();
+        let index = self.index() as isize;
+        unsafe {
+            self.set_len(len - 1);
+            self.set_index(index - 1);
+            Some(self.array.as_ref()[self.index()])
+        }
+    }
+
+    /// Returns an iterator over the elements as `&T`
+    #[inline]
+    pub fn iter(&self) -> Iter<T, A> {
```

Comment:
nit: I think it's better if this explicitly showed that `Iter` will freeze `Self` by explictly annotating the lifetimes: as in `fn iter<'a>(&'a self) -> Iter<'a, T, A>`.